File: frontend/src/game/board.js
/**
 * 게임 보드 모듈
 * 보드 상태 관리, 캐릭터 이동, 게임 로직을 구현합니다.
 */

import { Character, CharacterType, createFullTeam } from './character.js'
import { resolveBattle } from './rules.js'

export class Position {
  constructor(row, col) {
    this.row = row
    this.col = col
  }

  equals(other) {
    return this.row === other.row && this.col === other.col
  }
}

export class Move {
  constructor(fromPos, toPos) {
    this.fromPos = fromPos
    this.toPos = toPos
  }
}

export class GameBoard {
  static ROWS = 9
  static COLS = 7

  constructor() {
    // 보드: 2D 배열 (null 또는 Character)
    this.board = Array(GameBoard.ROWS)
      .fill(null)
      .map(() => Array(GameBoard.COLS).fill(null))

    // 플레이어 팀
    this.aiTeam = createFullTeam(0)
    this.playerTeam = createFullTeam(1)

    // 게임 상태
    this.currentPlayer = 1 // 1: Player, 0: AI
    this.gameOver = false
    this.winner = null

    // 전투 로그
    this.lastBattle = null
  }

  setupBoard(playerPositions = null) {
    // AI 캐릭터 배치 (상단 3x7 = 21칸)
    const aiPositions = Array.from({ length: 21 }, (_, i) => i)
    this.shuffle(aiPositions)

    let idx = 0
    for (let row = 0; row < 3; row++) {
      for (let col = 0; col < GameBoard.COLS; col++) {
        const charIdx = aiPositions[idx]
        this.board[row][col] = this.aiTeam[charIdx]
        idx++
      }
    }

    // Player 캐릭터 배치 (하단 3x7 = 21칸)
    if (!playerPositions) {
      playerPositions = Array.from({ length: 21 }, (_, i) => i)
      this.shuffle(playerPositions)
    }

    idx = 0
    for (let row = GameBoard.ROWS - 3; row < GameBoard.ROWS; row++) {
      for (let col = 0; col < GameBoard.COLS; col++) {
        const charIdx = playerPositions[idx]
        this.board[row][col] = this.playerTeam[charIdx]
        idx++
      }
    }
  }

  shuffle(array) {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1))
      ;[array[i], array[j]] = [array[j], array[i]]
    }
  }

  getCharacter(pos) {
    if (!this.isValidPosition(pos)) {
      return null
    }
    return this.board[pos.row][pos.col]
  }

  isValidPosition(pos) {
    return pos.row >= 0 && pos.row < GameBoard.ROWS && pos.col >= 0 && pos.col < GameBoard.COLS
  }

  getValidMoves(pos) {
    const character = this.getCharacter(pos)
    if (!character || !character.isAlive) {
      return []
    }

    if (!character.canMove) {
      return []
    }

    if (character.playerId !== this.currentPlayer) {
      return []
    }

    const validMoves = []
    const directions = [
      [-1, 0],
      [1, 0],
      [0, -1],
      [0, 1],
    ]

    for (const [dr, dc] of directions) {
      const newPos = new Position(pos.row + dr, pos.col + dc)

      if (!this.isValidPosition(newPos)) {
        continue
      }

      const target = this.getCharacter(newPos)

      // 빈 칸이거나, 상대방 캐릭터가 있는 경우만 이동 가능
      if (target === null || target.playerId !== character.playerId) {
        validMoves.push(new Move(pos, newPos))
      }
    }

    return validMoves
  }

  makeMove(move) {
    const character = this.getCharacter(move.fromPos)
    if (!character) {
      return false
    }

    // 유효한 이동인지 확인
    const validMoves = this.getValidMoves(move.fromPos)
    const isValid = validMoves.some(
      (m) => m.fromPos.equals(move.fromPos) && m.toPos.equals(move.toPos)
    )

    if (!isValid) {
      return false
    }

    const target = this.getCharacter(move.toPos)

    // 빈 칸으로 이동
    if (target === null) {
      this.board[move.toPos.row][move.toPos.col] = character
      this.board[move.fromPos.row][move.fromPos.col] = null
      this.lastBattle = null
    } else {
      // 전투 발생
      const battleResult = resolveBattle(character, target)

      // 전투 로그 저장
      this.lastBattle = {
        attacker: character.name,
        defender: target.name,
        attackerSurvives: battleResult.attackerSurvives,
        defenderSurvives: battleResult.defenderSurvives,
        attackerRevealed: battleResult.attackerRevealed,
        defenderRevealed: battleResult.defenderRevealed,
      }

      // 전투 결과 반영
      if (!battleResult.attackerSurvives) {
        character.isAlive = false
        this.board[move.fromPos.row][move.fromPos.col] = null
      }

      if (!battleResult.defenderSurvives) {
        target.isAlive = false
        this.board[move.toPos.row][move.toPos.col] = null
      }

      // 공격자가 승리하면 이동
      if (battleResult.attackerSurvives) {
        this.board[move.toPos.row][move.toPos.col] = character
        if (!move.fromPos.equals(move.toPos)) {
          this.board[move.fromPos.row][move.fromPos.col] = null
        }
      }

      // 캐릭터 공개 처리
      if (battleResult.attackerRevealed) {
        character.reveal()
      }
      if (battleResult.defenderRevealed) {
        target.reveal()
      }

      // 국기를 잡았는지 확인
      if (target.type === CharacterType.FLAG && !target.isAlive) {
        this.gameOver = true
        this.winner = character.playerId
      }
    }

    // 턴 변경
    this.currentPlayer = 1 - this.currentPlayer

    return true
  }

  getAllValidMoves(playerId) {
    const allMoves = []

    for (let row = 0; row < GameBoard.ROWS; row++) {
      for (let col = 0; col < GameBoard.COLS; col++) {
        const pos = new Position(row, col)
        const character = this.getCharacter(pos)

        if (character && character.isAlive && character.playerId === playerId) {
          const moves = this.getValidMoves(pos)
          for (const move of moves) {
            allMoves.push({ pos, move })
          }
        }
      }
    }

    return allMoves
  }

  clone() {
    const newBoard = new GameBoard()

    // 보드 복사
    newBoard.board = this.board.map((row) => [...row])

    // 팀 복사
    newBoard.aiTeam = [...this.aiTeam]
    newBoard.playerTeam = [...this.playerTeam]

    // 게임 상태 복사
    newBoard.currentPlayer = this.currentPlayer
    newBoard.gameOver = this.gameOver
    newBoard.winner = this.winner
    newBoard.lastBattle = this.lastBattle ? { ...this.lastBattle } : null

    return newBoard
  }

  getBoardState(forPlayer) {
    const state = []

    for (let row = 0; row < GameBoard.ROWS; row++) {
      const rowState = []
      for (let col = 0; col < GameBoard.COLS; col++) {
        const character = this.board[row][col]

        if (character === null) {
          rowState.push(null)
        } else {
          // 자기 편 캐릭터 또는 공개된 캐릭터는 정보 표시
          if (character.playerId === forPlayer || character.isRevealed) {
            rowState.push({
              name: character.name,
              type: character.typeName,
              player_id: character.playerId,
              is_alive: character.isAlive,
              is_revealed: character.isRevealed,
              priority: character.priority,
            })
          } else {
            // 상대방 캐릭터는 숨김
            rowState.push({
              name: '???',
              type: 'HIDDEN',
              player_id: character.playerId,
              is_alive: character.isAlive,
              is_revealed: false,
              priority: null,
            })
          }
        }
      }
      state.push(rowState)
    }

    return state
  }

  isGameOver() {
    return this.gameOver
  }

  getWinner() {
    return this.winner
  }
}

File: backend/game/board.py
"""
게임 보드 모듈
보드 상태 관리, 캐릭터 이동, 게임 로직을 구현합니다.
"""

import random
from typing import Optional, Tuple, List, Dict
from copy import deepcopy
from .character import Character, CharacterType, create_full_team
from .rules import resolve_battle, BattleResult


class Position:
    """보드 위치"""

    def __init__(self, row: int, col: int):
        self.row = row
        self.col = col

    def __eq__(self, other):
        if not isinstance(other, Position):
            return False
        return self.row == other.row and self.col == other.col

    def __hash__(self):
        return hash((self.row, self.col))

    def __repr__(self):
        return f"({self.row},{self.col})"


class Move:
    """이동 정보"""

    def __init__(self, from_pos: Position, to_pos: Position):
        self.from_pos = from_pos
        self.to_pos = to_pos

    def __repr__(self):
        return f"{self.from_pos} -> {self.to_pos}"


class GameBoard:
    """게임 보드"""

    ROWS = 9
    COLS = 7

    def __init__(self):
        """보드 초기화"""
        # 보드: 2D 배열 (None 또는 Character)
        self.board: List[List[Optional[Character]]] = [
            [None for _ in range(self.COLS)] for _ in range(self.ROWS)
        ]

        # 플레이어 팀
        self.ai_team: List[Character] = create_full_team(player_id=0)
        self.player_team: List[Character] = create_full_team(player_id=1)

        # 게임 상태
        self.current_player = 1  # 1: Player, 0: AI
        self.game_over = False
        self.winner: Optional[int] = None

        # 전투 로그
        self.last_battle: Optional[Dict] = None

    def setup_board(self, player_positions: Optional[List[int]] = None):
        """
        보드에 캐릭터 배치

        Args:
            player_positions: 플레이어 캐릭터 배치 순서 (0-20 인덱스 리스트)
                             None이면 자동 배치
        """
        # AI 캐릭터 배치 (상단 3x7 = 21칸)
        ai_positions = list(range(21))
        random.shuffle(ai_positions)

        idx = 0
        for row in range(3):
            for col in range(self.COLS):
                char_idx = ai_positions[idx]
                self.board[row][col] = self.ai_team[char_idx]
                idx += 1

        # Player 캐릭터 배치 (하단 3x7 = 21칸)
        if player_positions is None:
            player_positions = list(range(21))
            random.shuffle(player_positions)

        idx = 0
        for row in range(self.ROWS - 3, self.ROWS):
            for col in range(self.COLS):
                char_idx = player_positions[idx]
                self.board[row][col] = self.player_team[char_idx]
                idx += 1

    def get_character(self, pos: Position) -> Optional[Character]:
        """특정 위치의 캐릭터 가져오기"""
        if not self._is_valid_position(pos):
            return None
        return self.board[pos.row][pos.col]

    def _is_valid_position(self, pos: Position) -> bool:
        """유효한 위치인지 확인"""
        return 0 <= pos.row < self.ROWS and 0 <= pos.col < self.COLS

    def get_valid_moves(self, pos: Position) -> List[Move]:
        """
        특정 위치에서 가능한 모든 이동 반환

        Args:
            pos: 시작 위치

        Returns:
            가능한 Move 리스트
        """
        character = self.get_character(pos)
        if character is None or not character.is_alive:
            return []

        # 캐릭터가 움직일 수 없으면 (국기)
        if not character.can_move:
            return []

        # 현재 플레이어의 캐릭터가 아니면
        if character.player_id != self.current_player:
            return []

        valid_moves = []

        # 상하좌우 4방향
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        for dr, dc in directions:
            new_row = pos.row + dr
            new_col = pos.col + dc
            new_pos = Position(new_row, new_col)

            if not self._is_valid_position(new_pos):
                continue

            target = self.get_character(new_pos)

            # 빈 칸이거나, 상대방 캐릭터가 있는 경우만 이동 가능
            if target is None or target.player_id != character.player_id:
                valid_moves.append(Move(pos, new_pos))

        return valid_moves

    def make_move(self, move: Move) -> bool:
        """
        이동 실행

        Args:
            move: 이동 정보

        Returns:
            이동 성공 여부
        """
        character = self.get_character(move.from_pos)
        if character is None:
            return False

        # 유효한 이동인지 확인
        valid_moves = self.get_valid_moves(move.from_pos)
        is_valid = any(
            m.from_pos == move.from_pos and m.to_pos == move.to_pos
            for m in valid_moves
        )
        if not is_valid:
            return False

        target = self.get_character(move.to_pos)

        # 빈 칸으로 이동
        if target is None:
            self.board[move.to_pos.row][move.to_pos.col] = character
            self.board[move.from_pos.row][move.from_pos.col] = None
            self.last_battle = None
        else:
            # 전투 발생
            battle_result = resolve_battle(character, target)

            # 전투 로그 저장
            self.last_battle = {
                "attacker": character.name,
                "defender": target.name,
                "attacker_survives": battle_result.attacker_survives,
                "defender_survives": battle_result.defender_survives,
                "attacker_revealed": battle_result.attacker_revealed,
                "defender_revealed": battle_result.defender_revealed,
            }

            # 전투 결과 반영
            if not battle_result.attacker_survives:
                character.is_alive = False
                self.board[move.from_pos.row][move.from_pos.col] = None

            if not battle_result.defender_survives:
                target.is_alive = False
                self.board[move.to_pos.row][move.to_pos.col] = None

            # 공격자가 승리하면 이동
            if battle_result.attacker_survives:
                self.board[move.to_pos.row][move.to_pos.col] = character
                if move.from_pos != move.to_pos:
                    self.board[move.from_pos.row][move.from_pos.col] = None

            # 캐릭터 공개 처리
            if battle_result.attacker_revealed:
                character.reveal()
            if battle_result.defender_revealed:
                target.reveal()

            # 국기를 잡았는지 확인
            if target.type == CharacterType.FLAG and not target.is_alive:
                self.game_over = True
                self.winner = character.player_id

        # 턴 변경
        self.current_player = 1 - self.current_player

        return True

    def get_all_valid_moves(self, player_id: int) -> List[Tuple[Position, Move]]:
        """
        특정 플레이어의 모든 가능한 이동 반환

        Args:
            player_id: 플레이어 ID

        Returns:
            (시작 위치, Move) 튜플 리스트
        """
        all_moves = []

        for row in range(self.ROWS):
            for col in range(self.COLS):
                pos = Position(row, col)
                character = self.get_character(pos)

                if (
                    character is not None
                    and character.is_alive
                    and character.player_id == player_id
                ):
                    moves = self.get_valid_moves(pos)
                    for move in moves:
                        all_moves.append((pos, move))

        return all_moves

    def clone(self) -> "GameBoard":
        """보드 복사"""
        return deepcopy(self)

    def get_board_state(self, for_player: int) -> List[List[Optional[Dict]]]:
        """
        플레이어 시점에서 보드 상태 반환

        Args:
            for_player: 플레이어 ID (0: AI, 1: Player)

        Returns:
            2D 배열 (각 셀은 캐릭터 정보 딕셔너리 또는 None)
        """
        state = []

        for row in range(self.ROWS):
            row_state = []
            for col in range(self.COLS):
                character = self.board[row][col]

                if character is None:
                    row_state.append(None)
                else:
                    # 자기 편 캐릭터 또는 공개된 캐릭터는 정보 표시
                    if character.player_id == for_player or character.is_revealed:
                        row_state.append(
                            {
                                "name": character.name,
                                "type": character.type.name,
                                "player_id": character.player_id,
                                "is_alive": character.is_alive,
                                "is_revealed": character.is_revealed,
                                "priority": character.priority,
                            }
                        )
                    else:
                        # 상대방 캐릭터는 숨김
                        row_state.append(
                            {
                                "name": "???",
                                "type": "HIDDEN",
                                "player_id": character.player_id,
                                "is_alive": character.is_alive,
                                "is_revealed": False,
                                "priority": None,
                            }
                        )

            state.append(row_state)

        return state

    def is_game_over(self) -> bool:
        """게임 종료 확인"""
        return self.game_over

    def get_winner(self) -> Optional[int]:
        """승자 반환"""
        return self.winner
